step-api/src/main/java/org/jenkinsci/plugins/workflow/steps/AbstractSynchronousNonBlockingStepExecution.java

```diff
@@ -0,0 +1,84 @@
+package org.jenkinsci.plugins.workflow.steps;
+
+import hudson.util.DaemonThreadFactory;
+import hudson.util.NamingThreadFactory;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+/**
+ * Similar to {@link AbstractSynchronousStepExecution} but it always executes synchronously and does not block the CPS VM thread.
+ * @see {@link StepExecution}
+ * @param <T> the type of the return value (may be {@link Void})
+ */
+public abstract class AbstractSynchronousNonBlockingStepExecution<T> extends AbstractStepExecutionImpl {
+
+    private transient volatile Future<?> task;
+
+    protected AbstractSynchronousNonBlockingStepExecution() {
+    }
+
+    protected AbstractSynchronousNonBlockingStepExecution(StepContext context) {
+        super(context);
+    }
+
+    /**
+     * Meat of the execution.
+     *
+     * When this method returns, a step execution is over.
+     */
+    protected abstract T run() throws Exception;
+
+    @Override
+    public final boolean start() throws Exception {
+        task = StepRunnerExecutorService.get().submit(new StepRunner(this));
+        return false;
+    }
+
+    /**
+     * If the computation is going synchronously, try to cancel that.
+     */
+    @Override
+    public void stop(Throwable cause) throws Exception {
+        if (task != null) {
+            task.cancel(true);
+        }
+        getContext().onFailure(cause);
```

Comment:
Debatable whether it is right to call `onFailure` here (when `task != null`). It is more polite to let the `InterruptedException` be processed by `run` however it prefers—perhaps just throwing it up (in which case `StepRunner` will call `onFailure` anyway), but perhaps first doing some cleanup, maybe even deciding to return successfully if it turns out it no longer needed to wait.

See `DurableTaskStep.Execution.stop`. That tries to do a polite interruption the first time, but if you click the stop button a second time, assumes the process is unresponsive, gives up, and calls `onFailure` directly.

#141 currently interprets a _third_ abort attempt as a hard kill: the program is immediately discarded and the result set to `ABORTED` without any processing. Probably I should refine that so that the second abort attempt is handled in the infrastructure by calling `StepContext.onFailure` on the running step(s), removing the special logic in `DurableTaskStep.Execution`; in that case it would be safe to simply delete this line here.